### ISSUE TITLE
Missing dependencies when using array as `.CompilerInputPattern`

### ISSUE THREAD

**Issue by awefers** [fastbuild/fastbuild#1027]
I just tried upgrading from v1.11 to v1.12 (both for macOS and Windows). v1.12 introduces a linker error for the project that I am upgrading. The issue is related to collecting dependencies when using an array for a `Library`'s `.CompilerInputPattern` and traversing sub-directories.

The directory structure of the static library being compiled is as follows, and unlike all other source files in the project, this library uses `.cc` as extension for C++ source files:
```
src/directory_a/directory_b/directory_0/*.cc  // (to be excluded, not needed for the particular target)
src/directory_a/directory_b/directory_1/*.cc  // (to be excluded, not needed for the particular target)
src/directory_a/directory_b/directory_2/*.cc  // (to be excluded, not needed for the particular target)
src/directory_a/directory_b/*.cc
```
The library function is configured as follows:
```
.CompilerInputPattern      = { '*.cc', '*.cpp' } // actually inherrited via a Using() function
.CompilerInputPath         = '$ProjectPath$/src'
.CompilerInputPathRecurse  = true
.CompilerInputExcludePath  = { '$ProjectPath$/src/directory_a/directory_b/directory_0' } 
                           + { '$ProjectPath$/src/directory_a/directory_b/directory_1' }
                           + { '$ProjectPath$/src/directory_a/directory_b/directory_2' }
```
Using this settings will cause a linker error because none of the source files in `src/directory_a/directory_b/*.cc` are picked up as dependency and hence never get compiled. If I change
```
.CompilerInputPattern      = { '*.cc', '*.cpp' } 
```
to
```
.CompilerInputPattern      = '*.cc'
```
then the files in `src/directory_a/directory_b/*.cc` are picked up as dependency and the build succeeds.

Note:  To double check, I added `Print( .CompilerInputPattern )` inside the `Library` function when using the array and the output is as expected:
```
.CompilerInputPattern = // ArrayOfStrings, size: 2
{
    '*.cc'
    '*.cpp'
}
```

**Comment by ffulin** [comment]
I tried to reproduce the problem you've described without any luck (test case attached). The file in "directory_b" is found and the files in the "directory_[0|1|2]" sub-directories are ignored as expected. [Test.zip](https://github.com/fastbuild/fastbuild/files/14661887/Test.zip)

Maybe Ive misunderstood your example, or perhaps there is something more complex in your setup that is involved. Can you please take a look at the test case I made and compare it to your real-world case and let me know if anything stands out?

**Comment by awefers** [comment]
Hmm interesting, I just tried a standalone repro case by modifying your script and it does not reproduce either. The actual library that the dependencies fail on is ozz-animation. Here is a modified bff (with a `git clone https://github.com/guillaumeblanc/ozz-animation.git` right next to it), I am now wondering what else is causing this issue to  trigger. Let me try to reduce the actual build script so that will reproduce.
```
.ProjectPath = '.\ozz-animation'

Compiler( 'compiler' )
{
    .Executable                 = 'C:\Windows\System32\cmd.exe'
    .CompilerFamily             = 'custom'
}

.cpp =
[
    .CompilerInputPattern       = { '*.cc', '*.cpp' }
]

Library( 'all' )
{
    Using( .cpp )

    .Compiler                   = 'compiler'
    .CompilerOptions            = '/c echo %1 >> %2'

    .CompilerInputPath          = '$ProjectPath$/src'
    .CompilerInputPathRecurse   = true
    .CompilerInputExcludePath   = { '$ProjectPath$/src/animation/offline/fbx' } 
                                + { '$ProjectPath$/src/animation/offline/gltf' }
                                + { '$ProjectPath$/src/animation/offline/tools' } 
                                + { '$ProjectPath$/src/geometry' }
                                + { '$ProjectPath$/src/options' }
    .CompilerOutputPath         = '$ProjectPath$/out'

    .Librarian                  = '/usr/bin/touch'
    .LibrarianOptions           = '/c echo %1 >> %2'
    .LibrarianOutput            = '$ProjectPath$/out/final.txt'
}
```

**Comment by ffulin** [comment]
I think I may have found the problem.

In DirectoryListNode::FormatName:
```
    // Excluded paths
    if ( !excludeFiles.IsEmpty() ) <<----- This should be excludePaths
    {
        result += "|ePaths=";
        result.AppendList( excludePaths, '<' );
    }

    // Excluded files
    if ( !excludeFiles.IsEmpty() )
    {
        result += "|eFiles=";
        result.AppendList( excludeFiles, '<' );
    }
```

This would result in the excludedPaths not being correctly encoded in the name of the directory listing node. If there were multiple directory listings of the same path, with differing excluded paths, they would potentially be resolved to the same directly listing operation, causing incorrect results.

That would explain why:
a) my repro doesn't cause it (there is only 1 directory listing operation/no conflict)
b) removing *.cpp in your original case works (that uniquifies the nam as well)

I found this problem by running with -verbose and noticing that the excluded paths were not correctly encoded:
```
1> Dir: 'C:\test\CompilerInputPattern\src\|*.cc<*.cpp|r' (1 files)
 - C:\test\CompilerInputPattern\src\directory_a\directory_b\file.cc
```

**Comment by awefers** [comment]
Ah yes, that makes sense!

**Comment by awefers** [comment]
Thank you for taking a look!

**Comment by ffulin** [comment]
I've fixed the bug an added some test coverage: https://github.com/fastbuild/fastbuild/commit/bf8119a45e9602d368ceccb8f0d1c96206f6821f

That fix will be in the next release (v1.13). I'll leave this issue open until that version is released.

Thanks for reporting this bug!

**Comment by ffulin** [comment]
v1.13 has now been released and includes this fix. Thanks again for reporting the bug.
